### PR TITLE
[feat] Support locations in rules.

### PR DESCRIPTION
**Pull request by lukesneeringer**
~Uses #194 as a base.~ Merged.

This PR refactors how we think about locations.

1. It internalizes the distinction between a `lint.Location`
   and a `dpb.SourceCodeInfo_Location`. As far as rule authors
   are concerned, we only use the latter (as natively returned
   from `GetSourceInfo()`. Coversion happens when it is time
   to marshal to JSON or YAML. This avoids a situation where
   rule authors have to think about two different things with
   the same name.

2. It restores `Location` (but now the `dpb` one) as a first-class
   attribute on `Problem`.

3. It provides helpers for getting the package, syntax, or name.
   All helpers return the correct object to just be passed through.

### CODE REVIEW

**Comment by lukesneeringer** [comment]
I need to write more tests (specifically for `Problem`) before this is actually ready. I also need to run it and make sure it works. ;-)

File: lint/location.go [hunk 1]
```diff
@@ -14,43 +14,97 @@
...
 
-// Location describes a location in a source code file.
+// SyntaxLocation returns the location of the syntax definition in a file
+// descriptor.
 //
-// Note: positions are one-based.
-type Location struct {
-	Start Position `json:"start_position" yaml:"start_position"`
-	End   Position `json:"end_position" yaml:"end_position"`
+// If the location can not be found (for example, because there is no syntax
```
Comment (mingzhi) [inline lint/location.go]: Why not `nil`? Especially that we are returning a pointer.

File: lint/location.go [hunk 2]
```diff
@@ -14,43 +14,97 @@
...
-	Column int `json:"column_number" yaml:"column_number"`
+type sourceInfo map[string]*dpb.SourceCodeInfo_Location
+
+// findLocation returns the best location it can find for the given path.
+//
+// If the requested path can not be found, it climbs the ancestry tree until
+// it finds one, and ultimately returns a location corresponding to the
+// beginning of the file if it can not find anything.
+func (si sourceInfo) findLocation(path []int32) *dpb.SourceCodeInfo_Location {
+	// Base case: If we have no path left, send back a fake location that
```
Comment (mingzhi) [inline lint/location.go]: As we return a pointer, returning `nil` is more accurate.

File: lint/location.go [hunk 3]
```diff
@@ -14,43 +14,97 @@
...
+		}
+	}
+
+	// If the path exists in the source info registry, return that object.
+	if loc, ok := si[strPath(path)]; ok {
+		return loc
+	}
+
+	// No dice; try the immediate parent path.
+	return si.findLocation(path[:len(path)-1])
```
Comment (mingzhi) [inline lint/location.go]: If not found, we should return `nil`, and let developers decide what to do.

File: lint/location.go [hunk 4]
```diff
@@ -14,43 +14,97 @@
...
-// String returns the string representation.
-func (p Position) String() string {
-	return fmt.Sprintf("{line: %d, column: %d}", p.Line, p.Column)
+// sourceInfo compiles the source info object for a given file descriptor.
+// It also caches this into a registry, so subsequent calls using the same
+// descriptor will return the same object.
+func (sir sourceInfoRegistryType) sourceInfo(fd *desc.FileDescriptor) sourceInfo {
+	answer, ok := sir[fd]
+	if !ok {
+		answer = sourceInfo{}
```
Comment (mingzhi) [inline lint/location.go]: Should we use `make` to create a `map`?

File: lint/location.go [hunk 5]
```diff
@@ -14,43 +14,97 @@
...
+
+		// This file descriptor does not yet have a source info map.
+		// Compile one.
+		for _, loc := range fd.AsFileDescriptorProto().GetSourceCodeInfo().GetLocation() {
```
Comment (mingzhi) [inline lint/location.go]: We should check if source code infor is available, i.e., `GetSourceCodeInfo()` may return `nil`.

File: lint/location.go [hunk 6]
```diff
@@ -14,43 +14,97 @@
...
+		}
+
+		// Now that we calculated all of this, cache it on the registry so it
+		// does not need to be calculated again.
+		sir[fd] = answer
+	}
+	return answer
 }
+
+var sourceInfoRegistry = sourceInfoRegistryType{}
```
Comment (mingzhi) [inline lint/location.go]: Should we use `make` to create a `map`?

File: lint/location.go (hunk 1 above)
Comment (lukesneeringer) [reply lint/location.go]: Because this is about usefulness. The idea is that I _always_ want a valid location of some kind.

If someone sends a `Problem` and there is some unforeseen issue, the graceful fallback behavior is "put the error at the top of the file", as opposed to "return nil and probably panic later".

File: lint/location.go (hunk 3 above)
Comment (lukesneeringer) [reply lint/location.go]: Again, this is intentional, moving up the chain to the closest possible match.

The premise of "return `nil` and let developers decide what to do" does not hold water -- at the point that this does the wrong thing, it is out of the rule author's hands.

File: lint/location.go (hunk 4 above)
Comment (lukesneeringer) [reply lint/location.go]: Does it matter?

Comment (mingzhi) [reply lint/location.go]: My misunderstanding -- I thought map needs `make` to be initialized -- thanks!

File: lint/location.go (hunk 1 above)
Comment (mingzhi) [reply lint/location.go]: But in the `Problem`, we already do the fallback to get the location from the `Descriptor` if `Location == nil`, which in this case is the `FileDescriptor`.

Comment (lukesneeringer) [reply lint/location.go]: Right, and this is what we _use_ for that. It is therefore important that we return a valid location.

File: lint/location.go (hunk 3 above)
Comment (mingzhi) [reply lint/location.go]: Please note that `SourceCodeInfo_Location` has comments information. If, for example, a rule wants to check the comments of a package, it is wrong to return the file's comments (which may be still `nil`). But take one step back, I am also Okay with what you do here -- I don't have strong objection, and feel free to make it as-it and resolve our conservation. Thank you!

File: lint/location.go (hunk 1 above)
Comment (robcapo) [reply lint/location.go]: Agree that we should return nil. We can still provide convenience in the core linter (for example, as we already do: if locaiton is nil, we set it to the location of the descriptor).

If you truly want to have this fallback to the start location of the file, you can name the function something like `SyntaxLocationOrFileStart`, but I don't see the advantage of doing this.

The user will be using this to attach locations to problems. If we gracefully handle the `nil` case in the core linter code, we shouldn't have to worry about panics.

In what case do you see returning the start location as useful?

File: lint/problem.go [hunk 7]
```diff
@@ -15,34 +15,128 @@
...
+	return json.Marshal(p.marshal())
+}
+
+// MarshalYAML defines how to represent a Problem in YAML.
+func (p Problem) MarshalYAML() ([]byte, error) {
+	return yaml.Marshal(p.marshal())
+}
+
+// Marshal defines how to represent a serialized Problem.
+func (p Problem) marshal() interface{} {
```
Comment (robcapo) [inline lint/problem.go]: :+1: much happier with this than previous solution. Now we can also make `ruleId` and `category` private, rather than asking developers not to set them in comments.

File: lint/location.go (hunk 1 above)
Comment (lukesneeringer) [reply lint/location.go]: Convinced, will return `nil`. Waiting for #194 to be merged before making changes here.

File: lint/problem.go [hunk 8]
```diff
@@ -15,34 +15,128 @@
...
 	// Location provides the location of the problem.
 	// DO NOT SET: Set the descriptor instead.
-	Location Location `json:"location" yaml:"location"`
+	Location *dpb.SourceCodeInfo_Location
 
 	// RuleID provides the ID of the rule that this problem belongs to.
 	// DO NOT SET: this field will be set by the linter based on rule info
 	// and user configs.
-	RuleID RuleName `json:"rule_id" yaml:"rule_id"`
+	RuleID RuleName
```
Comment (robcapo) [inline lint/problem.go]: Make private, remove `DO NOT SET` docs

File: lint/problem.go [hunk 9]
```diff
@@ -15,34 +15,128 @@
...
 
 	// DO NOT SET:  this field will be set by the linter based on user configs.
-	Category string `json:"category,omitempty" yaml:"category,omitempty"`
+	Category string
```
Comment (robcapo) [inline lint/problem.go]: Make private, remove `DO NOT SET` docs

File: lint/problem.go (hunk 8 above)
Comment (lukesneeringer) [reply lint/problem.go]: Done.

File: lint/problem.go (hunk 9 above)
Comment (lukesneeringer) [reply lint/problem.go]: Agreed, done.

File: lint/location.go (hunk 5 above)
Comment (lukesneeringer) [reply lint/location.go]: This seems to work regardless.

File: lint/location.go [hunk 10]
```diff
@@ -14,43 +14,90 @@
...
-func (l Location) IsValid() bool {
-	return l.Start.IsValid() &&
-		l.End.IsValid() &&
-		(l.End.Line > l.Start.Line ||
-			l.End.Line == l.Start.Line && l.End.Column >= l.Start.Column)
+// PackageLocation returns the location of the package definition in a file
+// descriptor.
+//
+// If the location can not be found (for example, because there is no package
+// statement), it returns beginning of file.
```
Comment (mingzhi) [inline lint/location.go]: Update this comment. I think it now returns `nil` if not found.

File: lint/location.go (hunk 5 above)
Comment (mingzhi) [reply lint/location.go]: It seems that we always set `IncludeSourceCodeInfo` in the parser, so it works now. But for panic proof, we should check nil.

File: lint/location_test.go [hunk 11]
```diff
@@ -15,45 +15,113 @@
...
+				syntax = "proto3";
+
+				package google.api.linter;
+
+				message Foo {
+				  string bar = 1;
+				}
+			`)),
+		}),
+		IncludeSourceCodeInfo: true,
```
Comment (mingzhi) [inline lint/location_test.go]: Add a test by disabling `IncludeSourceCodeInfo`, and we should not panic, but return `nil` for locations.

File: lint/location.go (hunk 5 above)
Comment (mingzhi) [reply lint/location.go]: From your tests, it seems Okay. Thanks!

File: lint/location_test.go (hunk 11 above)
Comment (mingzhi) [reply lint/location_test.go]: Ignore it -- it seems you already have such tests.

File: lint/location.go (hunk 10 above)
Comment (lukesneeringer) [reply lint/location.go]: Fixed.
